odtp/dashboard/utils/middleware.py

```diff
@@ -0,0 +1,119 @@
+import jwt, logging
+from jwt import PyJWKClient
+from fastapi import Request, status, HTTPException, Cookie, Response
+from starlette.middleware.base import BaseHTTPMiddleware
+from nicegui import Client,app
+from odtp.dashboard.utils.storage import save_to_storage, get_from_storage
+from fastapi.responses import RedirectResponse, Response
+import odtp.mongodb.db as db
+
+
+unrestricted_page_routes = {'/', '/components'}
+
+def jwt_decode_header(encoded:str): 
+    """Decodes the header of a JWT token. (Giving data like algorithm and token type as well as signature key: alg, typ, kid.)"""
+    token_header = jwt.get_unverified_header(encoded)
+    return token_header
+
+def jwt_decode_payload(encoded: str):
+    """Decodes the payload of a JWT token without verifying JWT. (Giving data like issuer, groups, etc.)"""
+    payload = jwt.decode(jwt=encoded, algorithms=["RS256"], options={"verify_signature": False})
+    return payload
+
+def jwt_decode_from_client(encoded: str, url:str, audience:str):
+    """Decodes the payload of a JWT token using a client and verifying . (Giving data like issuer, groups, etc.)"""
+    print(url)
+    jwks_client = PyJWKClient(url)
+    signing_key = jwks_client.get_signing_key_from_jwt(encoded)
+    print(signing_key)
+    payload = jwt.decode(encoded, 
+                         signing_key.key, 
+                         audience=audience,
+                         algorithms=["RS256", "HS256"])
+    return payload
+
+def add_user( name_input, github_input, email_input, sub_input):
+    """ Add a new user if it doesn't exist and return the user ID."""
+    if db.user_exists(sub_input):
+        logging.warning(f"User with email {sub_input} already exists")
+        return None
+    else:
+        try:
+            user_id = db.add_user(
+                sub=sub_input,
+                name=name_input, 
+                github=github_input, 
+                email=email_input,
+                )
+            print(f"A user with id {user_id} has been created")
+        except Exception as e:
+            print(f"The user could not be added in the database. An Exception occurred: {e}")
+  
+
+class AuthMiddleware(BaseHTTPMiddleware):
+    def __init__(self, app, url, audience):
+        super().__init__(app)
+        self.url = url
+        self.audience = audience
+    async def dispatch(self, request:Request, call_next):
+        is_id_token_cookie_present = await self.is_cookie_empty_or_exists(request, "SWISSDATACUSTODIAN_IDTOKEN")
+        
+        """ Check if the token is in the browser otherwwise redirect to keycloack."""
+        
+        if not is_id_token_cookie_present:
+            if (
+                request.url.path in Client.page_routes.values() 
+                and request.url.path not in unrestricted_page_routes
+                ):
```

Comment:
I think it would be better to check if the `Authorization` header exists and contains a bearer token instead of the cookie, because we use the header and not the cookie